### PR TITLE
CONCF-94 beta Spinner in Lightbox galleries and VideoLoader change

### PR DESCRIPTION
**Pull request by dianafa**
https://wikia-inc.atlassian.net/browse/CONCF-94
https://wikia-inc.atlassian.net/browse/CONCF-211

PR combines solution proposed by Igor here: https://github.com/Wikia/mercury/tree/CONCF-94 for creating LoadingSpinnerMixin and my solution to lightbox galleries issues proposed here https://github.com/Wikia/mercury/pull/693. I they both combined together allow to difference lightbox with video and lightbox with image and move spinner control to their area of responsibility.

This PR contains also VideoLoader class refactorization as it should not affect HTML elements like it was before. Now this is maintained by VideoLightboxComponent.

ping @hakubo @rogatty 

this PR contains Mercury changes introduced here: https://github.com/Wikia/mercury/compare/HG-603-refactor-prototype-extensions

### CODE REVIEW

File: front/scripts/main/components/VideoLightboxComponent.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+'use strict';
+
+App.VideoLightboxComponent = Em.Component.extend({
+	classNames: ['lightbox-content-inner'],
+	classNameBindings: ['provider'],
+	videoLoader: null,
+
+	didInsertElement: function (): void {
+		this.initVideoPlayer();
+		this.set('provider', 'video-provider-' + this.videoLoader.getProviderName());
```
Comment (rogatty) [inline front/scripts/main/components/VideoLightboxComponent.ts]: Why do we set it manually if `provider` is a computed property? It duplicates line 21.

File: front/scripts/main/components/VideoLightboxComponent.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+	},
+
+	/**
+	 * @desc Computed property used to set class in template.
+	 * On the first launch this.videoLoader will not exist and it return ''.
+	 * As soon as the videoLoader will be set, the property will be changed.
+	 */
+	provider: function (): string {
+		if (this.videoLoader) {
```
Comment (rogatty) [inline front/scripts/main/components/VideoLightboxComponent.ts]: Always use getters for Ember.Object properties.
`var videoLoader = this.get('videoLoader');`

File: front/scripts/main/controllers/MediaLightboxController.ts [hunk 3]
```diff
@@ -97,7 +111,23 @@ App.MediaLightboxController = App.LightboxController.extend({
...
 	/**
-	 * gets current media from model
+	 * @desc checks if current media is a video or image
+	 * and which lightbox component to render
+	 *
+	 * @return string
+	 */
+	lightboxComponent: function (): string {
+		var currentMedia = this.get('currentMedia');
+		if (currentMedia && currentMedia.url) {
```
Comment (rogatty) [inline front/scripts/main/controllers/MediaLightboxController.ts]: Let's add a check if there is `currentMedia.type` too.

File: front/templates/main/components/video-lightbox.hbs [hunk 4]
```diff
@@ -0,0 +1,5 @@
+{{#if currentMedia}}
```
Comment (rogatty) [inline front/templates/main/components/video-lightbox.hbs]: Is it enough? Maybe we should check if `currentMedia.embed.html` exists?

**Comment by rogatty** [comment]
Should we remove `media-lightbox.hbs`?

File: front/templates/main/components/video-lightbox.hbs (hunk 4 above)
Comment (dianafa) [inline front/templates/main/components/video-lightbox.hbs]: It was enough so far :
https://github.com/Wikia/mercury/blob/dev/front/templates/main/media-lightbox.hbs#L1

Comment (rogatty) [inline front/templates/main/components/video-lightbox.hbs]: I know it was in the code but it doesn't mean it won't cause a bug someday :)

I leave it up to you if you want to check what will happen if `currentMedia` exists but `currentMedia.embed.html` does not.

**Comment by rogatty** [comment]
LGTM

File: front/scripts/main/components/ImageLightboxComponent.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+			this.load(currentMedia.url);
+		}
+	},
+
+	/**
+	 * @desc load an image and run update function when it is loaded
+	 *
+	 * @param url string - url of current image
+	 */
+	load: function(url: string): void {
```
Comment (kenkouot) [inline front/scripts/main/components/ImageLightboxComponent.ts]: space after function keyword

File: front/scripts/main/components/VideoLightboxComponent.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+	didInsertElement: function (): void {
+		this.initVideoPlayer();
+	},
+
+	/**
+	 * @desc Computed property used to set class in template.
+	 * On the first launch this.videoLoader will not exist and it return ''.
+	 * As soon as the videoLoader will be set, the property will be changed.
+	 */
+	provider: function (): string {
```
Comment (hakubo) [inline front/scripts/main/components/VideoLightboxComponent.ts]: as pre Ken's email
Em.computed

File: front/scripts/main/controllers/MediaLightboxController.ts [hunk 7]
```diff
@@ -32,7 +32,21 @@ App.MediaLightboxController = App.LightboxController.extend({
 	},
 
 	/**
-	 * This function checks if file=* matches any files on a page
+	 * @desc Checks if file from URL matches with currentMedia.
+	 * Handles situation when file is empty and when, 
+	 * 'back' button is pressed and other unexpected situations.
+	 */
+	fileObserver: function (): void {
```
Comment (hakubo) [inline front/scripts/main/controllers/MediaLightboxController.ts]: Em.observer

File: front/scripts/mercury/modules/VideoLoader.ts [hunk 8]
```diff
@@ -14,15 +14,11 @@ module Mercury.Modules {
...
 
 	export class VideoLoader {
-		element: HTMLElement;
 		data: any;
 		player: VideoPlayers.BasePlayer;
 
-		constructor (element: HTMLElement, data: any /* tracking cb */) {
-			element.innerHTML = data.html;
-			this.element = element;
+		constructor (data: any /* tracking cb */) {
```
Comment (hakubo) [inline front/scripts/mercury/modules/VideoLoader.ts]: is this comment still valid?
if yes please move it to the line above as per guidelines

**Comment by hakubo** [comment]
LGTM

**Comment by rogatty** [comment]
Jenkins didn't update the status but tests failed:
```
front/scripts/main/components/VideoLightboxComponent.ts(18,12): 2346 Supplied parameters do not match any signature of call target.
```

**Comment by rogatty** [comment]
LGTM

**Pull request merged**